dali/kernels/imgproc/convolution/convolution_gpu.h

```diff
@@ -0,0 +1,249 @@
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    DALI_ENFORCE(
+        in_shape.size() == window_size.size(),
+        make_string(
+            "Provided input shape and window sizes should have the same number of samples. Got: ",
+            in_shape.size(), " vs ", window_size.size(), "."));
+    int num_samples = in_shape.size();
+    for (int i = 0; i < num_samples; i++) {
+      int num_channels = has_channels ? in_shape[i][ndim - 1] : 1;
+      DALI_ENFORCE(
+          window_size[i][0] % 2 == 1,
+          make_string(
+              "Even or non-centered windows are not supported yet, got window with even length: ",
+              window_size, " for sample ", i, "."));
+
+      DALI_ENFORCE(window_size[i][0] * num_channels < kMaxWindowSize,
+                   make_string("Window is too big for sample ", i, ", got: ", window_size[i][0],
+                               ", expected at most: ", kMaxWindowSize / num_channels, "."));
+    }
+    se.add<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    se.add<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+    se.add<typename CutlassConv::SampleParams>(AllocType::GPU, num_samples);
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(in_shape);
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim> out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageCPU, const W, 1>& windows,
+           span<const int> window_anchors = {}, W scale = 1) {
+    int num_samples = in.size();
+
+    DALI_ENFORCE(
+        window_anchors.size() == num_samples || window_anchors.size() == 0,
+        make_string(
+            "Unexpected number of window_anchors, expected either anchors for all samples ( ",
+            num_samples,
+            ") or no anchors for windows centered by default, got: ", window_anchors.size(), "."));
+    auto* window_tmp_buffer_host_ptr =
+        ctx.scratchpad->Allocate<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    span<W> window_tmp_buffer_host(window_tmp_buffer_host_ptr, num_samples * kWindowCopyBufferSize);
+
+    // Pad and align windows in tmp memory, transfer the aligned windows to GPU
+    FillAlignedWindows(window_tmp_buffer_host, windows, in.shape);
+    // ToGPU allocates num_samples * kWindowCopyBufferSize * sizeof(W) memory through scratchpad
+    auto* window_tmp_buffer_gpu = ctx.scratchpad->ToGPU(ctx.gpu.stream, window_tmp_buffer_host);
+
+    Arguments args;
+    args.device_params_ptr =
+        ctx.scratchpad->Allocate<typename CutlassConv::SampleParams>(AllocType::GPU, num_samples);
+    if (kIsInnerConv) {
+      // Inner (innermost) - repack arguments
+      for (int i = 0; i < num_samples; i++) {
+        int window_size = static_cast<int>(windows.tensor_shape_span(i)[0]);
+        int window_anchor = window_anchors.size() ? window_anchors[i] : window_size / 2;
```

Comment:
```suggestion
        int window_anchor = windo_size - 1 - window_anchors.size() ? window_anchors[i] : window_size / 2;
```
assuming that we're reversing the windows and following the correlation formula.

Reply:
I'm not sure what the anchor means anymore.
0 would place the the window fully to the "left" of the input. I'm not sure if that's what we want. Probably renaming everything to the "correlation" would be a saner approach. I will put a big, fat TODO here to figure what we want to actually be.

Reply:
Anchor is the index in the window where the abscissa is 0.
```conv(f, g)(x) = sum(t) f(x - t) * g(t)```
where `g(t) = window[t - anchor]` if (t-anchor) is a valid index or 0 otherwise